MPN-933: complete exercises for Functional Programming in Scala, Chapter 5

yawaramin opened the pull request.

yawaramin committed:
laziness: implement and test Stream.toList

yawaramin committed:
laziness: implement and test Stream.take and Stream.drop

yawaramin committed:
laziness: implement and test Stream.takeWhile

yawaramin committed:
laziness: implement and test Stream.forAll

yawaramin committed:
laziness: implement and test Stream.headOption

yawaramin committed:
laziness: implement and test Stream.map, filter, append and flatMap

Commit:
laziness: implement and test Stream.constant

Commit:
laziness: implement and test Stream.from

The test for Stream.from revealed that Stream.take was reversing the
order of the returned values; fix this.

Commit:
laziness: implement and test Stream.fibs

Commit:
laziness: implement and test Stream.unfold and related functions

Write fibs, from, constant and ones in terms of unfold.

Commit:
laziness: implement and test more functions using Stream.unfold

Implement map, take, takeWhile, zipWith and zipAll. Test all of them.

Commit:
laziness: implement and test Stream.startsWith

Commit:
laziness: implement and test Stream.tails

Commit:
laziness: implement and test Stream.scanRight

jonas commented on exercises/src/main/scala/fpinscala/laziness/Stream.scala:
```diff
@@ -17,20 +18,80 @@ trait Stream[+A] {
-  def startsWith[B](s: Stream[B]): Boolean = sys.error("todo")
+  def flatMap[B](f: A => Stream[B]): Stream[B] =
+    foldRight(Empty: Stream[B]) { (a, b) =>
+      f(a).foldRight(b) { (a, b) => Cons(() => a, () => b) }
```
This can be simplified by using `append` (as suggested in the answers).

yawaramin replied:
@jonas thanks--I didn't use `append` at the time because in the sequence of working through the book, I hadn't implemented it at that point. But it does make things a lot simpler, so ... :-)

jonas commented on exercises/src/main/scala/fpinscala/laziness/Stream.scala:
```diff
@@ -17,20 +18,80 @@ trait Stream[+A] {
+  def tails: Stream[Stream[A]] =
+    unfold(this) { s =>
+      s match {
+        case Cons(h, t) => Option(s -> t())
```
Why use `Option` and not `Some` here? Can the expression be `null`?

yawaramin replied:
@jonas I've started using `Option` instead of `Some` everywhere to help Scala's type inference--`Option` is guaranteed to always give you a value of type `Option`, while `Some` may give you a value of type `Some`.

jonas commented on exercises/src/main/scala/fpinscala/laziness/Stream.scala:
```diff
@@ -17,20 +18,80 @@ trait Stream[+A] {
+  def zipAll[B](s2: Stream[B]): Stream[(Option[A], Option[B])] =
+    Stream.unfold(this -> s2) {
+      case (Cons(ha, ta), Cons(hb, tb)) =>
+        Option(Option(ha()) -> Option(hb()) -> (ta() -> tb()))
```
s/`Option`/`Some`/?

Commit:
laziness: simplify Stream.flatMap

As per @jonas comment at
https://github.com/yawaramin/fpinscala/pull/4#discussion_r28992679
